### ISSUE TITLE
AMDGPUAnnotateKernelFeatures should be deleted

### ISSUE THREAD

**Issue by arsenm** [llvm/llvm-project#63586]
This infers the "amdgpu-calls" and "amdgpu-stack-objects" attributes, which are poorly named and handled backwards. We only use these to try to infer whether we need to initialize flat scratch. Instead, we should use AMDGPUAttributor and infer a amdgpu-no-flat-scratch-init on kernels. We really want to detect whether any private addresses could potentially be addrspacecasted to flat.

**Comment by jdoerfert** [comment]
> private addresses could potentially be addrspacecasted to flat

So to detect if private addresses = AS(5) pointers(?) could be casted to AS(0)?

**Comment by arsenm** [comment]
> So to detect if private addresses = AS(5) pointers(?) could be casted to AS(0)?
Yes. Might as well consider AS(5) to anything and presume anything else is some future software used pointer tagging

**Comment by arsenm** [comment]
I think this is easy enough by extending the existing code in checkForQueuePtr. My main question is whether we should bother checking for constant expression addrspacecasts. The two possible cases are private globals (illegal and unhandled) and casts from integer literals (do we care about allowing assuming a 0 pointer points to something specific in the future?)

**Comment by nhaehnle** [comment]
inttoptr in general we want to support, I think. We do have some plans for using part of the wave scratch for non-stack purposes. It would then be a bit inconsistent to handle an inttoptr instruction but not an inttoptr constant expression.

**Comment by kzhuravl** [comment]
Need to re-assign to Jun.

**Comment by jwanggit86** [comment]
Is this just a matter of negating the adding of the AMDGPUAnnotateKernelFeature pass, or is there anything else required? Thanks!

**Comment by jwanggit86** [comment]
@arsenm The function `AMDGPUAnnotateKernelFeatures::addFeatureAttributes()` goes through the instructions to look for `alloca` and call instructions. However, there's a comment in the function that says this is a "poor way" to infer whether there are calls. So I wonder what's a better way to do this? Also what about stack objects, if not looking for `alloca`?

**Comment by arsenm** [comment]
> @arsenm The function `AMDGPUAnnotateKernelFeatures::addFeatureAttributes()` goes through the instructions to look for `alloca` and call instructions. However, there's a comment in the function that says this is a "poor way" to infer whether there are calls. So I wonder what's a better way to do this? Also what about stack objects, if not looking for `alloca`?

We don't actually care about calls, but addressable stack objects which may have their addresses captured by an addrspacecast. Looking for calls was only conservatively correct way to avoid knowing anything about the callee.

Attributor has checkForAllInstructions, which will easily let you find allocas. For the more sophisticated no-addrspacecast check, it will be a bit more effort (but the logic should approximately look the same as the NoCapture AA)

**Comment by jwanggit86** [comment]
@arsenm  
So if a function foo() has alloca, we set the attribute flat_scratch_init. Then all the callers of foo() should also have the attribute set. Is that how it's supposed to work?

**Comment by arsenm** [comment]
> @arsenm So if a function foo() has alloca, we set the attribute flat_scratch_init. Then all the callers of foo() should also have the attribute set. Is that how it's supposed to work?

Kind of, but it's inverted. If there cannot be an addrspacecasted alloca, (or more conservatively, if there cannot be an alloca), the function should set amdgpu-no-flat-scratch-init

**Comment by jwanggit86** [comment]
@arsenm 
So you want the attribute to be the negative, i.e. NO-flat-scratch-init? The current implementation in AnnotateKernelFeatures, appears to set the attributes for the positive cases.

**Comment by arsenm** [comment]
> @arsenm So you want the attribute to be the negative, i.e. NO-flat-scratch-init? The current implementation in AnnotateKernelFeatures, appears to set the attributes for the positive cases.

Correct, that's why AnnotateKernelFeatures should be deleted.

**Comment by jwanggit86** [comment]
@arsenm 
Do you want this `amdgpu-no-flat-scratch-init` to be part of `AAAMDAttributes`, or should it be separate?

**Comment by arsenm** [comment]
> @arsenm Do you want this `amdgpu-no-flat-scratch-init` to be part of `AAAMDAttributes`, or should it be separate?

It can go along with it. It's the same fundamental purpose, to trim out initialized user sgprs

**Comment by jwanggit86** [comment]
Created PR [94647](https://github.com/llvm/llvm-project/pull/94647)

**Comment by shiltian** [comment]
are we good enough to remove the pass now?

**Comment by arsenm** [comment]
> are we good enough to remove the pass now?

No, the follow up work to make use of the new attribute hasn't happened yet

**Comment by arsenm** [comment]
ping. @jwanggit86 are you working on the follow up work to make use of amdgpU-no-flat-scratch-init?

**Comment by jwanggit86** [comment]
> ping. [@jwanggit86](https://github.com/jwanggit86) are you working on the follow up work to make use of amdgpU-no-flat-scratch-init?

I'll work on it next week.
